[team-23(BE)] IssueTracker - 이슈 저장, 필터링 로직 수정

ak2j38 opened the pull request:
## 3주차 1번째 PR
- dan😁 안녕하세요! 로니와 아더입니다. 
- 3주차 첫 번째 PR을 보냅니다! 
- 남은 기간도 잘 부탁드립니다! 

## 진행 사항

### 1. 이슈 저장 및 조회 로직 구현 및 수정
  - 이슈 저장 로직 및 디테일 조회 로직, 필터링 로직을 구현했습니다.
  - 필터링을 위해 객체를 전달받는 대신 `id`를 전달받도록 수정했습니다.
  - 위의 변경사항에 따라 `queryDSL` 수정했습니다.

CMSSKKK committed:
refactor: 이슈 등록, 조회 관련해서 쿼리 및 로직, DTO 수정 및 추가

test: 이슈 등록 및 이슈 상세 조회 출력 테스트

CMSSKKK committed:
refactor: issue detail 조회 로직 리펙토링(reply 제외)

CMSSKKK committed:
feat: 이슈 디테일 조회 로직 완성

CMSSKKK committed:
test: 저장 및 조회 로직 test

CMSSKKK committed:
refactor: 이슈 저장 로직 cascade 활용 저장 및 Milestone, assignees, labels null checking

ak2j38 committed:
♻️ 2주차 2차 피드백 수정사항 적용

1. 애매한 메소드 네이밍 변경
2. 쓸모없는 주석 제거
3. @ResponseEntity 대신 @ResponseStatus 활용
4. 정적 팩토리 메소드 대신 새로운 메소드 적용
5. 멤버 반환 DTO에 id 추가

CMSSKKK committed:
Merge pull request #37 from ak2j38/be/refactor/feedback4

♻️ 2주차 2차 피드백 수정사항 적용

ak2j38 committed:
♻️ 이슈 필터링 작업 중

CMSSKKK committed:
refactor: issue-filtering 로직 간단 테스트 성공

CMSSKKK committed:
refactor: issue 필터링 로직 수정 완료

CMSSKKK committed:
fix: login argumentResolver 작동 설정

CMSSKKK committed:
refactor : api url mapping 및 로그인 필수 메서드 @Login 어노테이션 부착

CMSSKKK committed:
Merge pull request #39 from ak2j38/be/issue-filtering

Be/issue filtering

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/auth/AuthorizationExtractor.java:
```diff
@@ -1,6 +1,7 @@
 package com.ron2ader.issuetracker.auth;
 
 import org.springframework.http.HttpHeaders;
+import org.springframework.stereotype.Component;
```
불필요한 import를 정리하는 것도 클린 코드의 하나라고 생각해주세요.

CMSSKKK replied:
넵! 커밋전에 확인하는 습관 들이도록 하겠습니다!

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/controller/AuthController.java:
```diff
@@ -36,8 +36,8 @@ public LoginResponse requestAccessToken(String code) {
-    public Tokens requestNewTokens(@Login MemberDto memberDto) {
-        MemberDto findMember = memberService.findMember(memberDto.getMemberId());
+    public Tokens requestNewTokens(@Login String userId) {
+        MemberDto findMember = memberService.findMember(userId);
```
`nit` 
이 과정을 `LoginArgumentResolver`에서 하는 방법도 있을 것 같습니다.

CMSSKKK replied:
어떤 API에서는 `Member`를 db에서 조회할 필요가 없을 것이라고 생각해서 이렇게 설계했었는데, `argumentResolver` 활용에 대해서 더 공부해보겠습니다.

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/controller/LabelController.java:
```diff
@@ -30,21 +26,21 @@ public List<LabelsResponse> getLabels() {
 
     @PostMapping("/labels/{id}")
-    public ResponseEntity<LabelResponse> update(@PathVariable Long id, LabelRequest labelRequest) {
+    @ResponseStatus(HttpStatus.OK)
```
`200 OK`는 기본 값이기에 명시해주지 않아도 됩니다.

ak2j38 replied:
`200 OK`가 기본값인지 몰랐습니다. 😅
앞으로는 생략하겠습니다!

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/controller/issuedto/IssueDetail.java:
```diff
@@ -5,9 +5,11 @@
 
 @AllArgsConstructor
 @Getter
+@ToString
```
web 레이어의 dto라면 괜찮긴 하지만, 기본적으로 `ToString`의 활용을 권장하지 않습니다.

ak2j38 replied:
넵!  단순히 넘어오는 값 확인을 위해 추가했던 `ToString`인데 필요없는 코드들은 삭제하도록 하겠습니다.

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/controller/milestonedto/MilestoneResponse.java:
```diff
@@ -4,19 +4,23 @@
 
 @Getter
-@AllArgsConstructor(staticName = "of")
+@AllArgsConstructor
```
`nit` 정적 팩토리 메서드를 위한 생성자라면 접근 제한자를 설정하는 것을 권장합니다.

ak2j38 replied:
정적 팩토리 메소드를 위한 롬복 생성자라면 접근 제한자(`PROTECTED`, `PRIVATE`)를 꼭 추가하겠습니다.

Hyune-c replied:
~~사실 저도 실무에서는 누락하곤 합니다.~~

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/domain/issue/Issue.java:
```diff
@@ -34,17 +34,17 @@ public class Issue extends BaseEntity {
     private Milestone milestone;
 
-    @OneToMany(mappedBy = "issue", cascade = CascadeType.ALL, orphanRemoval = true)
+    @OneToMany(mappedBy = "issue", cascade = CascadeType.PERSIST, orphanRemoval = true)
```
cascade를 보수적으로 사용하는 것은 좋은 습관입니다. 👍

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/domain/issue/IssueRepository.java:
```diff
@@ -1,8 +1,23 @@
+            "left join fetch m.issues " +
+            "where i.id = :id")
+    @Override
+    Optional<Issue> findById(@Param("id") Long id);
```
`nit` 단순히 n+1을 막기 위함이라면 `@EntityGraph`를 활용하는 방법도 있습니다.
개인적으로는 jpql보다는 querydsl을 선호하긴 합니다.

ak2j38 replied:
N+1 문제를 해결하는 방법들에 대해서 학습해보았습니다.

1. `join fetch` 사용
2. `@EntityGraph` 사용

두 방법의 공통점은 필요한 컬렉션들을 `Eager`로 가져와 불필요한 N+1 쿼리를 방지한다는 것이고 차이점으로는 `join fetch`는 inner join을 사용하고 `@EntityGraph`의 경우에는 left outer join을 사용해 불필요한 row가 많이 생길 수 있다는 점입니다.

Hyune-c replied:
네. 그렇습니다. 심화 설명을 하자면

- `EntityGraph`를 사용하면 left outer join을 통해 불필요한 row가 생길 수 있는 것은 맞지만 
    - 먼저의 코멘트처럼 별도의 로직이 없이 단순히 n+1을 막기 위함이라면 고려할 수 있습니다.
    - 만약 그것으로 인해 성능이 문제가 간다고 하면 이미 이상한 설계이거나 다른 방법을 고려해야합니다.
    - `OneToFew` 개념에 대해 공부하시면 좋습니다.
- JPQL은 typesafe하지 않기에 개인적으로는 선호하지 않습니다.

Hyune-c commented on BE/src/test/java/com/ron2ader/issuetracker/service/IssueServiceTest.java:
```diff
@@ -1,76 +1,90 @@
+                1L);
+
+        // then
+        assertThat(createdIssueId).isEqualTo(4L);
```
SQL을 통해 생성된 issue가 3건 존재하기에 4라고 표현하셨군요..
SQL이 수정되면 영향을 받을 테스트인 것 같습니다.

CMSSKKK replied:
말씀대로 당장 통과만을 위한 테스트인 것 같습니다.
당장 이 테스트에서는 반환된 아이디로 다시 이슈를 찾아와서 값을 검증하는 것이 맞을까요??

그리고 data.sql의 역할은 applicaition이 실행할 때 기본적으로 들어가야할 데이터들을 저장하는 용도라고 다른 클래스원에게 들은 적이 있습니다.

test에서 data.sql을 사용하는 것에 대해서 어떻게 생각하시나요??

Hyune-c replied:
저라면 given 절에서 기존 issue개수를 체크하고, 예상되는 값을 그 값의 +1로 할 것 같아요.
이러면 SQL이 변경되어도 괜찮겠죠.

---

test에서 data.sql을 사용하는 것은 괜찮은 방법입니다.
Import 어노테이션을 활용할 수도 있구요.

하지만 운영용 sql과 tesst용 sql은 분리하는 것을 권장합니다.

Hyune-c commented on BE/src/test/java/com/ron2ader/issuetracker/domain/issue/IssueAssigneeRepositoryTest.java:
```diff
@@ -0,0 +1,34 @@
+        for (IssueAssignee issueAssignee : allByIssueId) {
+            System.out.println(issueAssignee.getAssignee());
+        }
+    }
```
단언문이 없는 테스트는 매우 이상합니다.

CMSSKKK replied:
잠시 쿼리 및 값을 확인해보려고 했었는데, 수정하지 못했던 것 같습니다. 수정하겠습니다!

Hyune-c commented on BE/src/main/java/com/ron2ader/issuetracker/service/MilestoneService.java:
```diff
@@ -25,39 +22,28 @@ public class MilestoneService {
-
-        return milestoneResponses;
+        return  milestones.stream().map(MilestoneResponse::from).collect(Collectors.toList());
     }
```
`nit` 
web 레이어의 response dto가 service 레이어에서 생성되고 있군요.
지금 규모에서는 나쁘지 않은 설계이지만 `레이어 역류`에 대해 알고 가시면 좋을 것 같습니다.

https://hyune.gitbook.io/hyune-wiki/and/codesquad/22reviewer/undefined/service-dto

CMSSKKK replied:
작성해주신 글 덕분에 dto와 레이어에 대해서 조금 더 이해하게 된 것 같습니다! 

설게를 진행할 때, 재사용성과 dto를 범용적으로 활용할 수 있는지에 대해서 더 고민해보도록 하겠습니다

Hyune-c reviewed (changes requested):
수고하셨습니다.

ak2j38 committed:
♻️ 3주차 1차 피드백 적용

1. 불필요한 import문 제거
2. LoginArgumentResolver에서 LoginMember 반환하도록 수정
3. 컨트롤러 @ResponseStatus 중 200 OK 제거
4. 불필요한 @ToString 제거
5. @AllArgsConstructor 접근제어자 추가
6. @EntityGraph 적용
7. 이슈등록 테스트코드 로직 수정
8. AuthInterceptor 추가
9. 불필요한 테스트코드 제거

CMSSKKK commented:
안녕하세요 단! 

아직 많이 부족한데... 내일 벌써 수료식이네요😅

불필요한 Import문이 너무 많아서 수정파일이 엄청 많이 생겼습니다.. 습관 잘 들이도록 하겠습니다. (수정사항을 보시기 불편할 것 같아서 걱정입니다.)

또한 `argumentResolver`만으로 로그인 검증이 필요한 api를 모두 커버하려면, member의 정보가 필요없는 api에도 파라미터로 `@Login`어노테이션을 부착해줘야하는 문제가 있어서 `interceptor`를 다시 추가하게되었습니다. 

마지막 PR이 남았지만, 감사인사 먼저 드립니다. 
3주간 친절하고 세심한 리뷰덕분에 많이 배웠습니다.
